Split construction, scanning, and building phases of S2WasmBuilder

dschuff opened the pull request:
Instead of doing all of the S2Wasm work in the constructor, split
construction, scanning (to determine implemented functions) and building
of the wasm module.

This allows the linker to get the symbol information (e.g. implemented
functions) without having to build an entire module (which will be
useful for archives) and to allow the linker to link a new object into
the existing one by building the wasm module in place on the existing
module.

kripken commented on src/asmjs/shared-constants.h:
```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+
+#ifndef wasm_shared_constants_h
```
based on the new location of this header, perhaps this name should include the subdir too?

dschuff replied:
Done.

kripken commented on src/s2wasm.h:
```diff
@@ -37,26 +37,41 @@ namespace wasm {
 
+  void build(LinkerObject *obj, LinkerObject::SymbolInfo* info) {
+    if (!obj->isEmpty()) Fatal() << "Cannot construct an S2WasmBuilder in an non-empty LinkerObject";
+    if (!info)
```
please `{}` or same line (same line sounds better to me)

dschuff replied:
Done.

kripken commented on src/wasm-linker.h:
```diff
@@ -44,6 +46,26 @@ class LinkerObject {
   };
+  // Information about symbols
+  struct SymbolInfo {
+    std::unordered_set<cashew::IString> implementedFunctions;
```
these should be `Name`

dschuff replied:
They actually can't be because it's `IString` and not `Name` that has `std::hash` defined for it. Originally I just thought it better to use IString as the key but now the hash maps are starting to proliferate, maybe I should revisit that. Anyway, I'd like to make that a separate PR.

kripken commented:
I landed 2 PRs that might cause merge conflicts here. I checked locally and the diff isn't that bad. I pushed it to `split-build-2` since I already did it anyhow, but feel free to ignore if it's not helpful.

dschuff commented:
OK, thanks for rebasing it. I pushed the commits from split-build-2 back onto split-build so I can keep using this PR. I think it looks right.

dschuff commented:
I addressed the comments (except for `Name` as noted above); I'm going to go ahead and land this to avoid more conflicts, as I expect to be out of the office until later this afternoon.
